### ISSUE TITLE
weird behavior under multithreading

### ISSUE THREAD

**Issue by jsenellart-systran** [torch/cutorch#586]
I am working on data parallelism for NMT and observed a wrong behaviour of some network when launched on different GPUs in different threads preventing completely their parallelisation.

The issue is reproduced in 2 minutes with the code below (need 4GPU cards - I am using GTX 1080, and at least 4 CPU).

I am comparing the execution time for launching the same computation totally independently, in different threads, on different devices - when launching 1 to 4 of these process. I expect the time to be constant since computations are launched in parallel.

This works well with network1 (built with `build_network1`) - time for 1 or 4 parallel computations is about the same (12 seconds).
This does not work with network2 (built with `build_network2` - just switch in the code) - time for 4 parallel computations is 4 times the time for one 1 computation - showing some bottleneck on a common shared resource (?). However, I don't see what could be this ressource considering that I am using 4 threads and 4 GPUs. It could be data transfer limitation between between GPU and CPU but there should not be any transfer for the computation since all data is on the GPU (?).

The computation graphs are slightly different in both networks, and network 2 is using nn.MM that network 1 is not.

The issue was initially observed when running the parallelism using cunn DataParallelTable - and narrowed down to the small independent code below

Any hint is welcome!
Thanks

```
local threads = require 'threads'

local timer=torch.Timer()

local function build_network1(opt)
  local inputs = {}
  table.insert(inputs, nn.Identity()())
  table.insert(inputs, nn.Identity()())

  local target_t = nn.Linear(opt.rnn_size, opt.rnn_size, false)(inputs[1]) -- batch_l x rnn_size
  local t = nn.Sum(3)(nn.Replicate(1,3)(target_t))
  local context = inputs[2] -- batch_l x source_timesteps x rnn_size

  attn = nn.Sum(2)(context)
  local context_combined = nn.JoinTable(2)({attn, t}) --  -- batch_l x rnn_size

  local context_output = nn.Linear(opt.rnn_size*2, opt.rnn_size, false)(context_combined)
  return nn.gModule(inputs, {context_output})
end

local function build_network2(opt)
  local inputs = {}
  table.insert(inputs, nn.Identity()())
  table.insert(inputs, nn.Identity()())

  local target_t = nn.Linear(opt.rnn_size, opt.rnn_size, false)(inputs[1]) -- batch_l x rnn_size
  local context = inputs[2] -- batch_l x source_timesteps x rnn_size

  local attn = nn.MM()({context, nn.Replicate(1,3)(target_t)}) -- batch_l x source_l x 1
  attn = nn.Sum(3)(attn)
  attn = nn.Replicate(1,2)(attn) -- batch_l x 1 x source_l
  local context_combined = nn.MM()({attn, context}) -- batch_l x 1 x rnn_size
  context_combined = nn.Sum(2)(context_combined) -- batch_l x rnn_size
  context_combined = nn.JoinTable(2)({context_combined, inputs[1]}) -- batch_l x rnn_size*2

  local context_output = nn.Linear(opt.rnn_size*2, opt.rnn_size, false)(context_combined)
  return nn.gModule(inputs, {context_output})
end

for i=1,4 do
  print('---- testing with ',i,' threads')
  timer:reset()
  pool=threads.Threads(
              i,
              function(threadid)
                print('start thread '..threadid)
                require 'cunn'
                require 'nngraph'
                cutorch.setDevice(threadid)

                n=build_network1({rnn_size=1000,dropout=0.2})
                print('\tbuilt network',threadid)
                n:cuda()
                input={torch.CudaTensor(50,1000):uniform(0.1),torch.CudaTensor(50,100,1000):uniform(0.1)}
                gradOutput=torch.CudaTensor(50,1000):uniform(1)
              end
  )
  pool:specific(true)
  pool:synchronize()
  timer:reset()
  for h=1,i do
    pool:addjob(h,function()
                  for i=1,10000 do
                    n:forward(input)
                    n:backward(input,gradOutput)
                  end
                end
    )
  end
  pool:synchronize()
  print(i,timer:time().real)
  pool:terminate()
end
```

**Comment by jsenellart-systran** [comment]
far simpler way to reproduce just using torch operations:

```
local threads = require 'threads'

for i=1,4 do
  pool=threads.Threads(
              i,
              function(threadid)
                require 'cutorch'
                cutorch.setDevice(threadid)
                a=torch.CudaTensor(50,100,1000):uniform(0.1)
                b=torch.CudaTensor(50,1000,1):uniform(0.1)
              end
  )
  pool:synchronize()

  local timer=torch.Timer()
  for h=1,i do
    pool:addjob(function()
                  for i=1,10000 do
                    torch.bmm(a,b)
                  end
                end
    )
  end
  pool:synchronize()
  print('total compute time with ' .. i .. ' threads:',timer:time().real)
  pool:terminate()
end
```

and the result is even worse:

```
total compute time with 1 threads:      5.1159970760345
total compute time with 2 threads:      9.575923204422
total compute time with 3 threads:      18.132717847824
total compute time with 4 threads:      35.368380069733 
```

When running 4 threads, I can see the 4 threads created and moderately working with the main thread running at 100% and GPUs at ~20%.

I must be missing something obvious but I don't see what... Thanks in advance!

**Comment by fmassa** [comment]
Maybe what you are seeing is due to the fact that current implementation of `torch.bmm` is a blocking operation in the GPU? Does it also happen with other operations that are non-blocking, like `torch.addmm` ?
Also, did you try using the new cache allocator that was recently made available? It could avoid the blocking operations on the GPU, making things faster. You can enable it with something like 
`THC_CACHE_ALLOCATOR=1 th myscript.lua`

**Comment by soumith** [comment]
As francisco said, and as mentioned in this section: https://github.com/torch/cutorch#cuda-memory-allocation

try this:

```
THC_CACHING_ALLOCATOR=1 th myscript.lua
```

**Comment by jsenellart-systran** [comment]
Thanks for your answers - if I correctly understand, operations creating some intermediate memory allocation will put some load on the main thread of the CPU which explains what I see.

@fmassa: you are right, I tried`M:addbmm(a,b)` and `M:baddbmm(a,b)` - the first one is behaving nicely (4s for 1-4 threads), the second is not (4s for 1 thread, 26s for 4 threads).

use of `THC_CACHING_ALLOCATOR=1` helps a little (4s for 1 thread, 24s for 4 threads).

is there any way to provide another memory allocation method? for my application, requested memory for each operation could be preallocated and reused.

**Comment by jsenellart-systran** [comment]
one easy way I found to workaround the problem is to replace

```
M:baddbmm(a,b)
```

by manual version of it:

```
for j=1,a:size(1) do
   M:select(1,j):addmm(a:select(1,j),b:select(1,j))
end
```

and it works - for the test above, I get now:
```
total compute time with 1 threads:      4.8067610263824 
total compute time with 2 threads:      4.8681681156158 
total compute time with 3 threads:      4.8319880962372 
total compute time with 4 threads:      5.0480268001556 
```

vs.

```
total compute time with 1 threads:      4.4713380336761 
total compute time with 2 threads:      8.4864900112152 
total compute time with 3 threads:      17.790560007095 
total compute time with 4 threads:      25.003059864044 
```

There is a small hit on performance for 1 thread - but it opens the data parallelism I was looking for - which means a x#GPU speed-up. Doing the same for the class MM works equally and my initial problem is fixed too.

@soumith - I looked at implementation of baddbmm in THCTensorMathBlas.cu - and from what I understand, the function creates CPU matrices for each batch, build them, and then allocate GPU matrices and transfer them to finally use `THCudaBlas_SgemmBatched`

```
  const real **matrices1 = (const real **)THAlloc(matrices_size);
[...]
  for (int i = 0; i < num_batches; ++i)
  {
    matrices1[i] = THCTensor_(data)(state, batch1_) + i * batch1_->stride[0];
[...]
  }
[...]
  const real **d_matrices1;
  THCudaCheck(THCudaMalloc(state, (void**)&d_matrices1, matrices_size));
[...]
  THCudaCheck(cudaMemcpyAsync(d_matrices1, matrices1, matrices_size,
                              cudaMemcpyHostToDevice, THCState_getCurrentStream(state)));
```

Is the cpu-allocation+gpu-allocation+transfer worth the possible optimization in THCudaBlas_DgemmBatched - could this be change to iteration on each batch to fix the problem at this level?

**Comment by soumith** [comment]
@jsenellart-systran thanks for the investigation. Tes, the cudaMemcpyAsync from HostToDevice is the problem, it is a sync point. We will discuss and try to work around this, i am not sure how exactly to solve this while keeping the API functional.

**Comment by jsenellart-systran** [comment]
For reference, I did test with the procedure described more nn modules (forward & backward) and here are the results:
- `Max`: OK
- `MM`: NOK (**7 times slower when running in 4 threads on 4 GPU**)
- `Add`: OK
- `CAdd`: NOK (**6 times slower**)
- `SoftMax`: OK
- `Linear`: OK
- `LogSoftMax`: OK
- `Reshape`: NOK (**1.5 times slower**)
- `Normalize`: OK

See [test2.lua.zip](https://github.com/torch/cutorch/files/578620/test2.lua.zip) for test code.

**Comment by soumith** [comment]
@jsenellart-systran the MM issue is being fixed by a colleague of mine.

for the rest of the modules, have you tried running with the caching allocator, i.e. ?
```
export THC_CACHING_ALLOCATOR=1
```

**Comment by jsenellart-systran** [comment]
@soumith - I did the test and you are right, except for MM, setting THC_CACHING_ALLOCATOR=1 fixes the other problem. Is there a reason this is not the default? Thanks for your support!

**Comment by soumith** [comment]
@jsenellart-systran it's not the default because it was recently implemented and is being tested. we didn't want to prematurely change the defaults. The MM fix will come in over the next few days, my colleague will convert the cpu logic into a cuda kernel and bypass the CPU-GPU transfer.

**Comment by soumith** [comment]
@jsenellart-systran a push has been fixed for MM as well. Now it wont be a sync point if using THC_CACHING_ALLOCATOR=1

**Comment by jsenellart-systran** [comment]
It works! tested with commit `7ed4888c337db9051bed4458e558733db5a8d48c` using `torch.bmm` and nn modules.

For nn modules: 

* default is without `THC_CACHING_ALLOCATOR`
* fix is with `THC_CACHING_ALLOCATOR=1` and latest patch

before patch, 'MM/4 threads fix' was the one remaining problematic.

| Module | 1 thread default | 4 threads default  | 1 thread fix | 4 threads fix  |
| --------- | --------------- | ------------------ | ----------- | ------------- |
| Max	| 1.883 | 1.900 | 1.890 | 1.910  |
| MM 	| 4.642 | **29.949** | _3.436_ | :boom: _3.464_ |
| Add	| 1.089 | 1.138 | 1.0649 | 1.100 |
| CAdd | **9.199** | **44.069** | _1.513_ | _1.528_ |
| Dropout	| 1.566 | 1.575 | 1.545 | 1.570 |
| SoftMax	| 1.394 | 1.408 | 1.393 | 1.410 |
| Linear | 0.876 | 0.904 | 0.875 | 0.887 |
| LogSoftMax | 0.526 | 0.539 | 0.528 | 0.539 |
| Reshape | 0.144 | 0.186 | 0.133 | 0.189 |
| Normalize | 2.128 | 2.232 | 2.077 | 2.111 |

Also, independently - use of `THC_CACHING_ALLOCATOR=1` improves speed for MM/CAdd.

Thanks @soumith and @1nadequacy !
